staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go

```diff
@@ -22,47 +22,59 @@ import (
-// information about the current user and true if successful, false if not successful,
-// or an error if the request could not be checked.
+// Request attempts to extract authentication information from a request and
+// returns a Response or an error if the request could not be checked.
 type Request interface {
-	AuthenticateRequest(req *http.Request) (user.Info, bool, error)
+	AuthenticateRequest(ctx context.Context, req *http.Request) (*Response, error)
 }
 
-// Password checks a username and password against a backing authentication store and
-// returns information about the user and true if successful, false if not successful,
-// or an error if the username and password could not be checked
+// Password checks a username and password against a backing authentication
+// store and returns a Response or an error if the password could not be
+// checked.
 type Password interface {
-	AuthenticatePassword(user, password string) (user.Info, bool, error)
+	AuthenticatePassword(ctx context.Context, user, password string) (*Response, error)
 }
 
 // TokenFunc is a function that implements the Token interface.
-type TokenFunc func(token string) (user.Info, bool, error)
+type TokenFunc func(ctx context.Context, token string) (*Response, error)
 
 // AuthenticateToken implements authenticator.Token.
-func (f TokenFunc) AuthenticateToken(token string) (user.Info, bool, error) {
+func (f TokenFunc) AuthenticateToken(ctx context.Context, token string) (*Response, error) {
 	return f(token)
 }
 
 // RequestFunc is a function that implements the Request interface.
-type RequestFunc func(req *http.Request) (user.Info, bool, error)
+type RequestFunc func(ctx context.Context, req *http.Request) (*Response, error)
 
 // AuthenticateRequest implements authenticator.Request.
-func (f RequestFunc) AuthenticateRequest(req *http.Request) (user.Info, bool, error) {
+func (f RequestFunc) AuthenticateRequest(ctx context.Context, req *http.Request) (*Response, error) {
 	return f(req)
 }
 
 // PasswordFunc is a function that implements the Password interface.
-type PasswordFunc func(user, password string) (user.Info, bool, error)
+type PasswordFunc func(ctx context.Context, user, password string) (*Response, error)
 
 // AuthenticatePassword implements authenticator.Password.
-func (f PasswordFunc) AuthenticatePassword(user, password string) (user.Info, bool, error) {
+func (f PasswordFunc) AuthenticatePassword(ctx context.Context, user, password string) (*Response, error) {
 	return f(user, password)
 }
+
+// Response is the struct returned by authenticator interfaces upon successful
+// authentication. It contains infomration about whether the authenticator
+// authenticated the request, information about the context of the
+// authentication, and information about the authenticated user.
+type Response struct {
+	// Audience is the audience selected by the authenticator. If the
+	// authenticator is not audience aware, this field will be nil.
+	Audience *string
```

Comment:
I still think this should be a slice:

```go
type Audiences []string

type Response struct {
	Audiences Audiences

	...
}
```

You can distinguish between `nil`, `len() == 0`, `len() > 0` as needed (though I am unsure if the distinction is required).

Reply:
Do you think it's valuable for an Authorizer to return multiple audiences that it was able to validate the token against? My idea wast that 1 valid audience is probably enough to return.

Reply:
I guess it depends on what question is being asked: "did any audience validate?" or "what audiences validated?"

If we are going with the former, would it simpler to just say `Audience *bool`?

`nil` == API server does not understand audiences
`true/false` == API server does understand audiences + validation was performed + the field holds the result